enhancements/oc/day2-add-nodes.md

```diff
@@ -0,0 +1,271 @@
+
+### Non-Goals
+- Replace MachineSet-based management of hosts for IPI clusters
+- Offer more or finer-grained configuration options for new nodes than those supported by the agent-based installer
+- Generation of the ISO image for adding the nodes into an environment not directly connected to the target cluster
+- Provide an automation for final nodes approval
+
+## Proposal
+A key factor for simplifying the user experience consists in offering a simple enough - but sufficiently flexible -  user interface 
+to accomplish the task of adding the required nodes to a cluster, and to monitor the progress of such activity. For this reason, 
+the proposal consists into enhancing the `oc` cli with two new commands:
+
+- `oc adm add-nodes-image` to generate the ISO for adding the nodes to the cluster
+- `oc adm monitor-add-nodes` to monitor the nodes currently being added to the cluster
+
+(More specific details about the proposed integration with `oc` could be find later in the [integration notes](#oc-integration-notes) paragraph)
+
+There are several reasons for choosing `oc` as a primary user interfaces for these tasks:
+- It’s the official cluster administration tool, so it is expected to be already available in the target cluster environment
+- It already supports the kubeconfig file for interacting with the target cluster (it’s relevant since it contains several pieces of 
+  information required for adding the nodes, as explained in the design section)
+- It already offers a couple of commands to manually approve/deny certificate signing requests (CSRs) when adding new nodes
+
+The usage of `oc` will also help in reducing the amount of information that the user will need to provide, resulting thus in a simplified
+input configuration file.
+
+### add-nodes-image command
+The format of the command will be the following:
+
+```
+oc --kubeconfig='' adm add-nodes-image [-o='agent-nodes.iso'] <nodes-config.yaml> 
+```
+
+Where:
+- `-o='agent-nodes.iso'`. The name of the ISO image to be generated.
+- `--kubeconfig=''`. Path to the kubeconfig file (or `KUBECONFIG` if available)
+
+`nodes-config.yaml` is a configuration file, similar to `agent-config.yaml`, used to specify the list of the hosts to be added as worker 
+nodes. For each host, it will be possible to specify a number of attributes, including a piece of nmstate configuration for the initial 
+static network configuration. If not specified, the tool will look for a `nodes-config.yaml` file in the current directory.
+
+The command will use the configured kubeconfig file, and the connection to the target cluster, to retrieve all the additional information 
+required to validate and generate the ISO image that will be used for adding the configured nodes. 
+
+#### nodes-config.yaml sample file
+Following an example of `nodes-config.yaml` file, where a couple of workers are described with an initial static network configuration:
+
+```
+hosts:
```

Comment:
Other inputs, including the install-config, look more like a kube API resource with version, group, kind, and other metadata. It would make sense to take that approach here, too, in case the schema of the input file needs to change in the future.

Is this structure definition going to be more or less the same as the one used in the installer for the same data? Can we share that code?

Reply:
> It would make sense to take that approach here, too, in case the schema of the input file needs to change in the future.

There's definitely a usability tradeoff here, in making the user include a bunch of boilerplate.
I'm not actually convinced the versioning does anything in this case. The thing that is effective is being disciplined about how you change fields. It's possible that using the k8s-style metadata acts as a reminder to developers about what can and cannot be safely changed, but it seems to me that's the main effect.

Versioning allows you to change the type of a field without changing the name of it. But that can be worked around by creating a new field with a different name - which you should probably do anyway. It also allows you to remove a field while still supporting old config files. But since you're still supporting the old field anyway there's no point creating a new version where the field is forbidden.

In practice the install-config version never changes and we just keep adding and deprecating fields without removing anything.